internal/redirecttraffic/redirect_traffic.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package redirecttraffic
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"strconv"
+
+	"github.com/hashicorp/consul-ecs/awsutil"
+	"github.com/hashicorp/consul-ecs/config"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/iptables"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	defaultProxyUserID      = 5995
+	defaultProxyInboundPort = 20000
+
+	consulDataplaneDNSBindHost = "127.0.0.1"
+	consulDataplaneDNSBindPort = 8600
+)
+
+type trafficRedirectProxyConfig struct {
+	BindPort           int    `mapstructure:"bind_port"`
+	PrometheusBindAddr string `mapstructure:"envoy_prometheus_bind_addr"`
+	StatsBindAddr      string `mapstructure:"envoy_stats_bind_addr"`
+}
+
+type TrafficRedirectionCfg struct {
+	ProxySvc *api.AgentService
+
+	ConsulServerAddress  string
+	ClusterARN           string
+	ProxyHealthCheckPort int
+
+	EnableConsulDNS      bool
+	ExcludeInboundPorts  []int
+	ExcludeOutboundPorts []int
+	ExcludeOutboundCIDRs []string
+	ExcludeUIDs          []string
+
+	iptablesCfg iptables.Config
+
+	// Fields used only for unit tests
+	iptablesProvider iptables.Provider
+}
+
+type TrafficRedirectionProvider interface {
+	Apply() error
+
+	// Used to fetch the resultant config in unit tests
+	config() iptables.Config
+}
+
+type TrafficRedirectionOpts func(*TrafficRedirectionCfg)
+
+func WithIPTablesProvider(provider iptables.Provider) TrafficRedirectionOpts {
+	return func(c *TrafficRedirectionCfg) {
+		c.iptablesProvider = provider
+	}
+}
+
+func New(cfg *config.Config, proxySvc *api.AgentService, consulServerAddress, clusterARN string, proxyHealthCheckPort int, opts ...TrafficRedirectionOpts) TrafficRedirectionProvider {
+	input := &TrafficRedirectionCfg{
+		ProxySvc:             proxySvc,
+		ConsulServerAddress:  consulServerAddress,
+		ClusterARN:           clusterARN,
+		ProxyHealthCheckPort: proxyHealthCheckPort,
+		EnableConsulDNS:      cfg.ConsulDNSEnabled(),
+		ExcludeInboundPorts:  cfg.TransparentProxy.ExcludeInboundPorts,
+		ExcludeOutboundPorts: cfg.TransparentProxy.ExcludeOutboundPorts,
+		ExcludeOutboundCIDRs: cfg.TransparentProxy.ExcludeOutboundCIDRs,
+		ExcludeUIDs:          cfg.TransparentProxy.ExcludeUIDs,
+	}
+
+	for _, opt := range opts {
+		opt(input)
```

Comment:
This is a way to provide optional function arguments in Go. When someone invokes this function in [this](https://github.com/hashicorp/consul-ecs/blob/7eba201cc0b4852c57324ad611a224956fd2a6e8/internal/redirecttraffic/redirect_traffic_test.go#L210) manner, the corresponding `WithIPTablesProvider` handler gets called which alters the input config to add the passed provider.

This loop is meant to add multiple such optional parameters in the future if needed. For example, let's say there is a new optional field `foo` that gets added in the struct, it can be populated with the help of a function that looks something like

```golang
func WithFoo(foo string) TrafficRedirectionOpts {
	return func(c *TrafficRedirectionCfg) {
		c.foo = foo
	}
}
```

and the caller would do something like

```golang
trafficRedirectionProvider.Apply(
.....
.....
WithIPTablesProvider(customIPTablesProvider),
WithFoo("foo"),
)
```

This loop will iterator through the `opts` input (which are nothing but functions) and start applying them with `opt(input)`